File: frontend/src/lib/geofences.ts
import { api } from '@/lib/api';

export interface Geofence {
  id: string;
  name: string;
  category: string | null;
  centerLat: number;
  centerLng: number;
  radiusM: number;
  active: boolean;
  createdAt: Date;
  updatedAt: Date;
}

export type GeofenceEventType = 'enter' | 'exit';

export interface GeofenceEvent {
  id: string;
  geofenceId: string;
  truckId: string;
  event: GeofenceEventType;
  latitude: number;
  longitude: number;
  recordedAt: Date;
}

interface BackendGeofence {
  id: string;
  name: string;
  category: string | null;
  center_lat: number;
  center_lng: number;
  radius_m: number;
  active: boolean;
  created_at: string;
  updated_at: string;
}

interface BackendGeofenceEvent {
  id: string;
  geofence_id: string;
  truck_id: string;
  event: GeofenceEventType;
  latitude: number;
  longitude: number;
  recorded_at: string;
}

function adapt(g: BackendGeofence): Geofence {
  return {
    id: g.id,
    name: g.name,
    category: g.category,
    centerLat: Number(g.center_lat),
    centerLng: Number(g.center_lng),
    radiusM: Number(g.radius_m),
    active: g.active,
    createdAt: new Date(g.created_at),
    updatedAt: new Date(g.updated_at),
  };
}

function adaptEvent(e: BackendGeofenceEvent): GeofenceEvent {
  return {
    id: e.id,
    geofenceId: e.geofence_id,
    truckId: e.truck_id,
    event: e.event,
    latitude: Number(e.latitude),
    longitude: Number(e.longitude),
    recordedAt: new Date(e.recorded_at),
  };
}

export interface GeofenceInput {
  name: string;
  category?: string;
  centerLat: number;
  centerLng: number;
  radiusM: number;
  active?: boolean;
}

function toBody(input: Partial<GeofenceInput>): Record<string, unknown> {
  const body: Record<string, unknown> = {};
  if (input.name !== undefined) body.name = input.name;
  if (input.category !== undefined) body.category = input.category || null;
  if (input.centerLat !== undefined) body.center_lat = input.centerLat;
  if (input.centerLng !== undefined) body.center_lng = input.centerLng;
  if (input.radiusM !== undefined) body.radius_m = input.radiusM;
  if (input.active !== undefined) body.active = input.active;
  return body;
}

export const geofencesApi = {
  list: async (): Promise<Geofence[]> => {
    const data = await api<BackendGeofence[]>('/api/geofences');
    return data.map(adapt);
  },
  create: async (input: GeofenceInput): Promise<Geofence> => {
    const data = await api<BackendGeofence>('/api/geofences', { method: 'POST', body: toBody(input) });
    return adapt(data);
  },
  update: async (id: string, patch: Partial<GeofenceInput>): Promise<Geofence> => {
    const data = await api<BackendGeofence>(`/api/geofences/${id}`, { method: 'PUT', body: toBody(patch) });
    return adapt(data);
  },
  remove: async (id: string): Promise<void> => {
    await api<void>(`/api/geofences/${id}`, { method: 'DELETE' });
  },
  events: async (limit = 50): Promise<GeofenceEvent[]> => {
    const data = await api<BackendGeofenceEvent[]>(`/api/geofences/events?limit=${limit}`);
    return data.map(adaptEvent);
  },
};

File: backend/app/models/geofences.py
from __future__ import annotations

import uuid
from datetime import datetime, timezone

from sqlalchemy import String, Boolean, DateTime, Numeric, Enum, ForeignKey, Index
from sqlalchemy.dialects.postgresql import UUID
from sqlalchemy.orm import Mapped, mapped_column, relationship

from app.core.database import Base
from app.models.enums import GeofenceEventType


class Geofence(Base):
    """A circular geofence. Enter/exit transitions are detected on GPS ingest.

    Circular (center + radius) keeps evaluation a cheap haversine check with no
    PostGIS dependency — works identically on Postgres and the SQLite test DB.
    """

    __tablename__ = "geofences"

    id: Mapped[uuid.UUID] = mapped_column(UUID(as_uuid=True), primary_key=True, default=uuid.uuid4)
    org_id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True), ForeignKey("organizations.id", ondelete="CASCADE"), nullable=False, index=True
    )
    name: Mapped[str] = mapped_column(String(100), nullable=False)
    category: Mapped[str | None] = mapped_column(String(40), nullable=True)  # depot, customer, restricted, ...
    center_lat: Mapped[float] = mapped_column(Numeric(10, 8), nullable=False)
    center_lng: Mapped[float] = mapped_column(Numeric(11, 8), nullable=False)
    radius_m: Mapped[float] = mapped_column(Numeric(10, 2), nullable=False)  # meters
    active: Mapped[bool] = mapped_column(Boolean, default=True, nullable=False)
    created_at: Mapped[datetime] = mapped_column(DateTime(timezone=True), default=lambda: datetime.now(timezone.utc), nullable=False)
    updated_at: Mapped[datetime] = mapped_column(DateTime(timezone=True), default=lambda: datetime.now(timezone.utc), nullable=False)

    events: Mapped[list["GeofenceEvent"]] = relationship(back_populates="geofence", cascade="all, delete-orphan")


class GeofenceEvent(Base):
    """A truck crossing the boundary of a geofence (enter or exit)."""

    __tablename__ = "geofence_events"
    __table_args__ = (
        Index("ix_geofence_events_truck_geofence", "truck_id", "geofence_id", "recorded_at"),
    )

    id: Mapped[uuid.UUID] = mapped_column(UUID(as_uuid=True), primary_key=True, default=uuid.uuid4)
    geofence_id: Mapped[uuid.UUID] = mapped_column(UUID(as_uuid=True), ForeignKey("geofences.id", ondelete="CASCADE"), nullable=False)
    truck_id: Mapped[uuid.UUID] = mapped_column(UUID(as_uuid=True), ForeignKey("trucks.id", ondelete="CASCADE"), nullable=False)
    event: Mapped[GeofenceEventType] = mapped_column(Enum(GeofenceEventType, name="geofence_event_type"), nullable=False)
    latitude: Mapped[float] = mapped_column(Numeric(10, 8), nullable=False)
    longitude: Mapped[float] = mapped_column(Numeric(11, 8), nullable=False)
    recorded_at: Mapped[datetime] = mapped_column(DateTime(timezone=True), default=lambda: datetime.now(timezone.utc), nullable=False)

    geofence: Mapped["Geofence"] = relationship(back_populates="events")
